File: src/characterCodes.ts
export enum CharacterCode {
  Blank = 0,
  A = 1,
  B = 2,
  C = 3,
  D = 4,
  E = 5,
  F = 6,
  G = 7,
  H = 8,
  I = 9,
  J = 10,
  K = 11,
  L = 12,
  M = 13,
  N = 14,
  O = 15,
  P = 16,
  Q = 17,
  R = 18,
  S = 19,
  T = 20,
  U = 21,
  V = 22,
  W = 23,
  X = 24,
  Y = 25,
  Z = 26,
  One = 27,
  Two = 28,
  Three = 29,
  Four = 30,
  Five = 31,
  Six = 32,
  Seven = 33,
  Eight = 34,
  Nine = 35,
  Zero = 36,
  ExclamationMark = 37,
  AtSign = 38,
  PoundSign = 39,
  DollarSign = 40,
  LeftParen = 41,
  RightParen = 42,
  Hyphen = 44,
  PlusSign = 46,
  Ampersand = 47,
  EqualsSign = 48,
  Semicolon = 49,
  Colon = 50,
  SingleQuote = 52,
  DoubleQuote = 53,
  PercentSign = 54,
  Comma = 55,
  Period = 56,
  Slash = 59,
  QuestionMark = 60,
  DegreeSign = 62,
  Red = 63,
  Orange = 64,
  Yellow = 65,
  Green = 66,
  Blue = 67,
  Violet = 68,
  White = 69,
  Black = 70,
  Filled = 71,
}

const validCharacterCodes = Object.keys(CharacterCode).map(
  (key) => CharacterCode[key]
);

const CHARACTER_CODES = [
  {
    code: CharacterCode.Blank,
    name: "Blank",
    mappings: [
      " ",
      "©",
      "®",
      "<",
      ">",
      "²",
      "†",
      "‡",
      "ˆ",
      "Þ",
      "þ",
      "µ",
      "¶",
      "*",
      "^",
      "_",
      "¬",
      "«",
      "»",
      "›",
      "³",
      "¹",
      "€",
      "‹",
      "˜",
      "÷",
    ],
  },
  {
    code: CharacterCode.A,
    name: "A",
    mappings: [
      "A",
      "a",
      "â",
      "à",
      "å",
      "á",
      "À",
      "Á",
      "Â",
      "Ã",
      "Å",
      "ã",
      "ä",
      "Ä",
      "æ",
      "Æ",
    ],
  },
  {
    code: CharacterCode.B,
    name: "B",
    mappings: ["B", "b", "ß"],
  },
  {
    code: CharacterCode.C,
    name: "C",
    mappings: ["C", "c", "ç", "Ç", "¢", "ć", "Ć", "č", "Č"],
  },
  {
    code: CharacterCode.D,
    name: "D",
    mappings: ["D", "d", "Ð", "ð"],
  },
  {
    code: CharacterCode.E,
    name: "E",
    mappings: ["E", "e", "é", "ê", "ë", "è", "È", "É", "Ê", "Ë", "€", "£"],
  },
  {
    code: CharacterCode.F,
    name: "F",
    mappings: ["F", "f", "ƒ", "ſ"],
  },
  {
    code: CharacterCode.G,
    name: "G",
    mappings: ["G", "g", "ğ", "Ğ", "ģ", "Ģ", "ġ", "Ġ", "ĝ", "Ĝ"],
  },
  {
    code: CharacterCode.H,
    name: "H",
    mappings: ["H", "h", "ħ", "Ħ", "ĥ", "Ĥ"],
  },
  {
    code: CharacterCode.I,
    name: "I",
    mappings: ["I", "i", "í", "ï", "î", "ì", "Ì", "Í", "Î", "Ï", "|", "¡"],
  },
  {
    code: CharacterCode.J,
    name: "J",
    mappings: ["J", "j", "ĵ", "Ĵ", "į", "Į"],
  },
  {
    code: CharacterCode.K,
    name: "K",
    mappings: ["K", "k", "ķ", "Ķ", "ĸ"],
  },
  {
    code: CharacterCode.L,
    name: "L",
    mappings: ["L", "l", "£", "ł", "Ł", "ļ", "Ļ", "ĺ", "Ĺ", "ľ", "Ľ", "ŀ", "Ŀ"],
  },
  {
    code: CharacterCode.M,
    name: "M",
    mappings: ["M", "m"],
  },
  {
    code: CharacterCode.N,
    name: "N",
    mappings: ["N", "n", "ñ", "Ñ", "ń", "Ń", "ň", "Ň", "ņ", "Ņ"],
  },
  {
    code: CharacterCode.O,
    name: "O",
    mappings: [
      "O",
      "o",
      "ó",
      "ô",
      "ò",
      "Ò",
      "Ó",
      "Ô",
      "Õ",
      "Ø",
      "ð",
      "õ",
      "ø",
      "ö",
      "Ö",
      "œ",
      "Œ",
    ],
  },
  {
    code: CharacterCode.P,
    name: "P",
    mappings: ["P", "p", "Þ", "þ", "¶"],
  },
  {
    code: CharacterCode.Q,
    name: "Q",
    mappings: ["Q", "q"],
  },
  {
    code: CharacterCode.R,
    name: "R",
    mappings: ["R", "r", "ŕ", "Ŕ", "ř", "Ř", "ŗ", "Ŗ"],
  },
  {
    code: CharacterCode.S,
    name: "S",
    mappings: ["S", "s", "š", "Š", "§", "ś", "Ś", "ş", "Ş", "ș", "Ș"],
  },
  {
    code: CharacterCode.T,
    name: "T",
    mappings: ["T", "t", "ť", "Ť", "ţ", "Ţ", "ŧ", "Ŧ"],
  },
  {
    code: CharacterCode.U,
    name: "U",
    mappings: [
      "U",
      "u",
      "û",
      "ù",
      "ú",
      "Ù",
      "Ú",
      "Û",
      "Ü",
      "ü",
      "µ",
      "ū",
      "Ū",
      "ů",
      "Ů",
      "ų",
      "Ų",
    ],
  },
  {
    code: CharacterCode.V,
    name: "V",
    mappings: ["V", "v", "Ʋ", "ʋ"],
  },
  {
    code: CharacterCode.W,
    name: "W",
    mappings: ["W", "w", "ŵ", "Ŵ", "ẁ", "Ẁ", "ẃ", "Ẃ", "ẅ", "Ẅ"],
  },
  {
    code: CharacterCode.X,
    name: "X",
    mappings: ["X", "x", "ẍ", "Ẍ"],
  },
  {
    code: CharacterCode.Y,
    name: "Y",
    mappings: ["Y", "y", "ý", "ÿ", "Ý", "ŷ", "Ŷ", "ỳ", "Ỳ", "ỹ", "Ỹ", "Ÿ"],
  },
  {
    code: CharacterCode.Z,
    name: "Z",
    mappings: ["Z", "z", "ž", "Ž", "ź", "Ź", "ż", "Ż"],
  },
  {
    code: CharacterCode.One,
    name: "1",
    mappings: ["1", "¹"],
  },
  {
    code: CharacterCode.Two,
    name: "Two",
    mappings: ["2", "²"],
  },
  {
    code: CharacterCode.Three,
    name: "Three",
    mappings: ["3", "³"],
  },
  {
    code: CharacterCode.Four,
    name: "Four",
    mappings: ["4"],
  },
  {
    code: CharacterCode.Five,
    name: "Five",
    mappings: ["5"],
  },
  {
    code: CharacterCode.Six,
    name: "Six",
    mappings: ["6"],
  },
  {
    code: CharacterCode.Seven,
    name: "Seven",
    mappings: ["7"],
  },
  {
    code: CharacterCode.Eight,
    name: "Eight",
    mappings: ["8"],
  },
  {
    code: CharacterCode.Nine,
    name: "Nine",
    mappings: ["9"],
  },
  {
    code: CharacterCode.Zero,
    name: "Zero",
    mappings: ["0", "º"],
  },
  {
    code: CharacterCode.ExclamationMark,
    name: "ExclamationMark",
    mappings: ["!", "¡", "ǃ"],
  },
  {
    code: CharacterCode.AtSign,
    name: "AtSign",
    mappings: ["@"],
  },
  {
    code: CharacterCode.PoundSign,
    name: "PoundSign",
    mappings: ["#", "№"],
  },
  {
    code: CharacterCode.DollarSign,
    name: "DollarSign",
    mappings: ["$", "¢", "£", "¤", "¥", "₩", "₪", "₫", "€", "₹", "₺", "₽"],
  },
  {
    code: CharacterCode.LeftParen,
    name: "LeftParen",
    mappings: ["(", "[", "{", "⟨", "«"],
  },
  {
    code: CharacterCode.RightParen,
    name: "RightParen",
    mappings: [")", "]", "}", "⟩", "»"],
  },
  {
    code: CharacterCode.Hyphen,
    name: "Hyphen",
    mappings: ["-", "—", "–", "¯", "~"],
  },
  {
    code: CharacterCode.PlusSign,
    name: "PlusSign",
    mappings: ["+", "±", "∓", "∔"],
  },
  {
    code: CharacterCode.Ampersand,
    name: "Ampersand",
    mappings: ["&"],
  },
  {
    code: CharacterCode.EqualsSign,
    name: "EqualsSign",
    mappings: ["="],
  },
  {
    code: CharacterCode.Semicolon,
    name: "Semicolon",
    mappings: [";", ";", "；"],
  },
  {
    code: CharacterCode.Colon,
    name: "Colon",
    mappings: [":", "¦"],
  },
  {
    code: CharacterCode.SingleQuote,
    name: "SingleQuote",
    mappings: [
      "'",
      "‘",
      "’",
      "`",
      "´",
      "‚",
      "‛",
      "ʹ",
      "ʻ",
      "ʽ",
      "ʾ",
      "ʿ",
      "ˈ",
      "ˊ",
      "ˋ",
    ],
  },
  {
    code: CharacterCode.DoubleQuote,
    name: "DoubleQuote",
    mappings: [`"`, "„", "“", "”", "¨", "˝", "ˮ", "˵", "˶"],
  },
  {
    code: CharacterCode.PercentSign,
    name: "PercentSign",
    mappings: ["%", "‰", "‱"],
  },
  {
    code: CharacterCode.Comma,
    name: "Comma",
    mappings: [",", "¸", "‚", "，", "、", "､"],
  },
  {
    code: CharacterCode.Period,
    name: "Period",
    mappings: [".", "․", "‥", "…"],
  },
  {
    code: CharacterCode.Slash,
    name: "Slash",
    mappings: ["/", "\\", "⁄", "∕", "⧸", "⫻", "⫽", "⧵"],
  },
  {
    code: CharacterCode.QuestionMark,
    name: "QuestionMark",
    mappings: ["?", "¿"],
  },
  {
    code: CharacterCode.DegreeSign,
    name: "DegreeSign",
    mappings: ["°", "¤", "•", "·", "∙", "∘", "⚬", "⦿", "⨀", "⨁", "⨂"],
  },
];

const _supportedCharacters = [
  ...CHARACTER_CODES.map((characterCode) =>
    characterCode.mappings[0]?.toLocaleLowerCase()
  ),
  ...CHARACTER_CODES.map((characterCode) =>
    characterCode.mappings[0]?.toUpperCase()
  ),
  ...["\n", "“", "‘", "⬜", "🟥", "🟧", "🟨", "🟩", "🟦", "🟪", "⬛"],
];

export const supportedCharacters = _supportedCharacters.filter(
  (item, index) => _supportedCharacters.indexOf(item) === index
);

const MAPPED_CHARACTERS = CHARACTER_CODES.reduce((prev, current) => {
  return {
    ...prev,
    ...current.mappings.reduce((prevMapping, currentMapping) => {
      return {
        ...prevMapping,
        [currentMapping]: current.code,
      };
    }, {}),
  };
}, {});

export const getCharacterCode = (character: string) =>
  MAPPED_CHARACTERS[character];

const validateCharacterCode = (code: number) => {
  if (validCharacterCodes.includes(code)) {
    return code as CharacterCode;
  }

  throw new Error(`Invalid Character Code: ${code}`);
};

export const convertCharactersToCharacterCodes = (characters: string) =>
  characters.split("").reduce(
    (prev: any, current, index) => {
      const characterCode = prev.isCharacterCode
        ? characters[index + 1] === "}"
          ? validateCharacterCode(+current)
          : validateCharacterCode(+`${current}${characters[index + 1]}`)
        : null;

      return {
        accumulator:
          current === "{" || current === "}" || prev.skipNext
            ? prev.accumulator
            : [
                ...prev.accumulator,
                prev.isCharacterCode
                  ? characterCode
                  : getCharacterCode(current),
              ],
        isCharacterCode: current === "{",
        skipNext: prev.isCharacterCode && characters[index + 1] !== "}",
      };
    },
    {
      accumulator: [],
      isCharacterCode: false,
      skipNext: false,
    }
  ).accumulator;

export const mappingToCharacter = (character: string) => {
  if (supportedCharacters.includes(character)) {
    return character;
  }

  const characterCodeObject = CHARACTER_CODES.find((code) =>
    code.mappings.includes(character)
  );

  return characterCodeObject?.mappings[0]?.toLocaleLowerCase() || " ";
};

File: python/pyvbml/character_codes.py
"""Character codes.

Port of Vestaboard/vbml/src/characterCodes.ts
"""

from __future__ import annotations

from enum import IntEnum
from typing import Final

from .multiple_character_mappings import MULTIPLE_CHARACTER_MAPPINGS

# __VARIATION SELECTOR-16__
# https://www.unicode.org/charts/PDF/UFE00.pdf
# Emoji variation selector; preceded by U+2764 ("Heavy Black Heart") to produce
# "Red Heart" (❤️). Should be removed entirely, not cast to whitespace.
_VARIATION_SELECTOR_SIXTEEN = "\ufe0f"


class CharacterCode(IntEnum):
    """Character code."""

    BLANK = 0
    LETTER_A = 1
    LETTER_B = 2
    LETTER_C = 3
    LETTER_D = 4
    LETTER_E = 5
    LETTER_F = 6
    LETTER_G = 7
    LETTER_H = 8
    LETTER_I = 9
    LETTER_J = 10
    LETTER_K = 11
    LETTER_L = 12
    LETTER_M = 13
    LETTER_N = 14
    LETTER_O = 15
    LETTER_P = 16
    LETTER_Q = 17
    LETTER_R = 18
    LETTER_S = 19
    LETTER_T = 20
    LETTER_U = 21
    LETTER_V = 22
    LETTER_W = 23
    LETTER_X = 24
    LETTER_Y = 25
    LETTER_Z = 26
    ONE = 27
    TWO = 28
    THREE = 29
    FOUR = 30
    FIVE = 31
    SIX = 32
    SEVEN = 33
    EIGHT = 34
    NINE = 35
    ZERO = 36
    EXCLAMATION_MARK = 37
    AT_SIGN = 38
    POUND_SIGN = 39
    DOLLAR_SIGN = 40
    LEFT_PARENTHESIS = 41
    RIGHT_PARENTHESIS = 42
    HYPHEN = 44
    PLUS_SIGN = 46
    AMPERSAND = 47
    EQUALS_SIGN = 48
    SEMICOLON = 49
    COLON = 50
    SINGLE_QUOTE = 52
    DOUBLE_QUOTE = 53
    PERCENT_SIGN = 54
    COMMA = 55
    PERIOD = 56
    SLASH = 59
    QUESTION_MARK = 60
    DEGREE_SIGN = 62
    RED = 63
    ORANGE = 64
    YELLOW = 65
    GREEN = 66
    BLUE = 67
    VIOLET = 68
    WHITE = 69
    BLACK = 70
    FILLED = 71


COLOR_CODES: Final[list[CharacterCode]] = [
    CharacterCode.RED,
    CharacterCode.ORANGE,
    CharacterCode.YELLOW,
    CharacterCode.GREEN,
    CharacterCode.BLUE,
    CharacterCode.VIOLET,
    CharacterCode.WHITE,
    CharacterCode.BLACK,
]

_valid_character_codes = set(int(c) for c in CharacterCode)

_CHARACTER_CODES_TABLE = [
    (
        CharacterCode.BLANK,
        [
            " ",
            "©",
            "®",
            "<",
            ">",
            "²",
            "†",
            "‡",
            "ˆ",
            "Þ",
            "þ",
            "µ",
            "¶",
            "*",
            "^",
            "¬",
            "«",
            "»",
            "›",
            "³",
            "¹",
            "€",
            "‹",
            "˜",
            "÷",
            "π",
            "∆",
            "√",
            "∫",
            "∞",
        ],
    ),
    (
        CharacterCode.LETTER_A,
        [
            "A",
            "a",
            "â",
            "à",
            "å",
            "á",
            "À",
            "Á",
            "Â",
            "Ã",
            "Å",
            "ã",
            "ä",
            "Ä",
            "∂",
            "œ",
            "æ",
            "Æ",
        ],
    ),
    (CharacterCode.LETTER_B, ["B", "b"]),
    (CharacterCode.LETTER_C, ["C", "c", "ç", "Ç", "¢", "ć", "Ć", "č", "Č"]),
    (CharacterCode.LETTER_D, ["D", "d", "Ð", "ð"]),
    (
        CharacterCode.LETTER_E,
        ["E", "e", "é", "ê", "ë", "è", "È", "É", "Ê", "Ë", "€", "£", "∑"],
    ),
    (CharacterCode.LETTER_F, ["F", "f", "ƒ", "ſ"]),
    (CharacterCode.LETTER_G, ["G", "g", "ğ", "Ğ", "ģ", "Ģ", "ġ", "Ġ", "ĝ", "Ĝ"]),
    (CharacterCode.LETTER_H, ["H", "h", "ħ", "Ħ", "ĥ", "Ĥ"]),
    (
        CharacterCode.LETTER_I,
        ["I", "i", "í", "ï", "î", "ì", "Ì", "Í", "Î", "Ï", "|", "¡"],
    ),
    (CharacterCode.LETTER_J, ["J", "j", "ĵ", "Ĵ", "į", "Į"]),
    (CharacterCode.LETTER_K, ["K", "k", "ķ", "Ķ", "ĸ"]),
    (
        CharacterCode.LETTER_L,
        ["L", "l", "£", "ł", "Ł", "ļ", "Ļ", "ĺ", "Ĺ", "ľ", "Ľ", "ŀ", "Ŀ"],
    ),
    (CharacterCode.LETTER_M, ["M", "m"]),
    (CharacterCode.LETTER_N, ["N", "n", "ñ", "Ñ", "ń", "Ń", "ň", "Ň", "ņ", "Ņ"]),
    (
        CharacterCode.LETTER_O,
        ["O", "o", "ó", "ô", "ò", "Ò", "Ó", "Ô", "Õ", "Ø", "ð", "õ", "ø", "ö", "Ö"],
    ),
    (CharacterCode.LETTER_P, ["P", "p", "Þ", "þ", "¶"]),
    (CharacterCode.LETTER_Q, ["Q", "q"]),
    (CharacterCode.LETTER_R, ["R", "r", "ŕ", "Ŕ", "ř", "Ř", "ŗ", "Ŗ"]),
    (CharacterCode.LETTER_S, ["S", "s", "š", "Š", "§", "ś", "Ś", "ş", "Ş", "ș", "Ș"]),
    (CharacterCode.LETTER_T, ["T", "t", "ť", "Ť", "ţ", "Ţ", "ŧ", "Ŧ"]),
    (
        CharacterCode.LETTER_U,
        [
            "U",
            "u",
            "û",
            "ù",
            "ú",
            "Ù",
            "Ú",
            "Û",
            "µ",
            "ū",
            "Ū",
            "ů",
            "Ů",
            "ų",
            "Ų",
            "Ü",
        ],
    ),
    (CharacterCode.LETTER_V, ["V", "v", "Ʋ", "ʋ"]),
    (CharacterCode.LETTER_W, ["W", "w", "ŵ", "Ŵ", "ẁ", "Ẁ", "ẃ", "Ẃ", "ẅ", "Ẅ"]),
    (CharacterCode.LETTER_X, ["X", "x", "ẍ", "Ẍ"]),
    (
        CharacterCode.LETTER_Y,
        ["Y", "y", "ý", "ÿ", "Ý", "ŷ", "Ŷ", "ỳ", "Ỳ", "ỹ", "Ỹ", "Ÿ"],
    ),
    (CharacterCode.LETTER_Z, ["Z", "z", "ž", "Ž", "ź", "Ź", "ż", "Ż"]),
    (CharacterCode.ONE, ["1", "¹"]),
    (CharacterCode.TWO, ["2", "²"]),
    (CharacterCode.THREE, ["3", "³"]),
    (CharacterCode.FOUR, ["4"]),
    (CharacterCode.FIVE, ["5"]),
    (CharacterCode.SIX, ["6"]),
    (CharacterCode.SEVEN, ["7"]),
    (CharacterCode.EIGHT, ["8"]),
    (CharacterCode.NINE, ["9"]),
    (CharacterCode.ZERO, ["0"]),
    (CharacterCode.EXCLAMATION_MARK, ["!", "ǃ"]),
    (CharacterCode.AT_SIGN, ["@"]),
    (CharacterCode.POUND_SIGN, ["#", "№"]),
    (
        CharacterCode.DOLLAR_SIGN,
        ["$", "¢", "£", "¤", "¥", "₩", "₪", "₫", "€", "₹", "₺", "₽"],
    ),
    (CharacterCode.LEFT_PARENTHESIS, ["(", "[", "{", "⟨", "«"]),
    (CharacterCode.RIGHT_PARENTHESIS, [")", "]", "}", "⟩", "»"]),
    (CharacterCode.HYPHEN, ["-", "—", "–", "¯", "~", "_"]),
    (CharacterCode.PLUS_SIGN, ["+", "±", "∓", "∔"]),
    (CharacterCode.AMPERSAND, ["&"]),
    (CharacterCode.EQUALS_SIGN, ["=", "≠", "≈", "≡"]),
    (CharacterCode.SEMICOLON, [";", "；"]),
    (CharacterCode.COLON, [":", "¦"]),
    (
        CharacterCode.SINGLE_QUOTE,
        ["'", "'", "'", "`", "´", "‚", "‛", "ʹ", "ʻ", "ʽ", "ʾ", "ʿ", "ˈ", "ˊ", "ˋ"],
    ),
    (
        CharacterCode.DOUBLE_QUOTE,
        ['"', "„", "\u201c", "\u201d", "¨", "˝", "ˮ", "˵", "˶", "‟", "\u201f"],
    ),
    (CharacterCode.PERCENT_SIGN, ["%", "‰", "‱"]),
    (CharacterCode.COMMA, [",", "¸", "‚", "，", "、", "､"]),
    (CharacterCode.PERIOD, [".", "․", "‥", "…"]),
    (CharacterCode.SLASH, ["/", "\\", "⁄", "∕", "⧸", "⫻", "⫽", "⧵"]),
    (CharacterCode.QUESTION_MARK, ["?", "¿"]),
    (
        CharacterCode.DEGREE_SIGN,
        [
            "°",
            "˚",
            "º",
            "¤",
            "•",
            "·",
            "∙",
            "∘",
            "⚬",
            "⦿",
            "⨀",
            "⨁",
            "⨂",
            "❤️",
            "🧡",
            "💛",
            "💚",
            "💙",
            "💜",
            "🖤",
            "🤍",
            "🤎",
            "❤",
        ],
    ),
]

_supported: list[str] = []
for _code, _mappings in _CHARACTER_CODES_TABLE:
    if _mappings:
        _supported.append(_mappings[0].lower())
        _supported.append(_mappings[0].upper())

_supported += [
    "\n",
    "\u201c",
    "\u2019",
    "{",
    "}",
    "⬜",
    "🟥",
    "🟧",
    "🟨",
    "🟩",
    "🟦",
    "🟪",
    "⬛",
    "❤",
]

supported_characters: list[str] = list(
    dict.fromkeys(_supported)
)  # deduplicated, order preserved

# Flat char → code lookup (first mapping wins)
_MAPPED_CHARACTERS: dict[str, int] = {}
for _code, _mappings in _CHARACTER_CODES_TABLE:
    for _m in _mappings:
        if _m not in _MAPPED_CHARACTERS:
            _MAPPED_CHARACTERS[_m] = int(_code)


def get_character_code(character: str) -> int | None:
    """Get character code."""
    return _MAPPED_CHARACTERS.get(character)


def _validate_character_code(code: int) -> int:
    """Validate character code."""
    if code in _valid_character_codes:
        return code
    raise ValueError(f"Invalid Character Code: {code}")


def convert_characters_to_character_codes(characters: str) -> list[int]:
    """Convert a string (possibly containing {N} colour/character codes) into a list of integer character codes."""
    result: list[int] = []
    i = 0
    while i < len(characters):
        ch = characters[i]
        if ch == "{":
            j = characters.index("}", i)
            code_str = characters[i + 1 : j]
            result.append(_validate_character_code(int(code_str)))
            i = j + 1
        elif ch == "}":
            i += 1
        else:
            code = get_character_code(ch)
            if code is not None:
                result.append(code)
            i += 1
    return result


def mapping_to_character(character: str) -> str:
    """Map a character."""
    if character == _VARIATION_SELECTOR_SIXTEEN:
        return ""

    multi = MULTIPLE_CHARACTER_MAPPINGS.get(character)
    if multi:
        return multi

    if character in supported_characters:
        return character

    for _code, mappings in _CHARACTER_CODES_TABLE:
        if character in mappings:
            return mappings[0].lower()

    return " "
